File: NCoreUtils.Logging/LoggerScope.fs
namespace NCoreUtils.Logging

open System
open System.Threading

[<Sealed>]
type internal LoggerScope (stack : (obj list) ref, index : int) =
  let mutable isDisposed = 0
  interface IDisposable with
    member __.Dispose () =
      if 0 = Interlocked.CompareExchange (&isDisposed, 1, 0) then
        let mutable success = false
        while not success do
          let s = !stack
          match s with
          | _ :: s' when List.length s' >= index -> Interlocked.CompareExchange(stack, s', s) |> ignore
          | _ -> success <- true

File: NCoreUtils.Logging/Internal/LoggerScope.cs
using System;
using System.Threading;

namespace NCoreUtils.Logging.Internal;

internal class LoggerScope(Logger logger, int index) : IDisposable
{
    private readonly Logger _logger = logger ?? throw new ArgumentNullException(nameof(logger));

    private readonly int _index = index;

    private int _isDisposed;

    public void Dispose()
    {
        if (0 == Interlocked.CompareExchange(ref _isDisposed, 1, 0))
        {
            bool success;
            do
            {
                var stack = _logger._stack.Root;
                if (Scope.Count(stack) >= _index)
                {
                    success = ReferenceEquals(stack, _logger._stack.CompareExchange(Scope.Truncate(stack, _index), stack));
                }
                else
                {
                    success = true;
                }
            }
            while (!success);
        }
    }
}